clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1739,17 +1738,57 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     InitExprs.push_back(MemberInit.get());
   }
 
+  int getDims() {
+    int Dims = 0;
+    for (int i = MemberExprBases.size() - 1; i >= 0; --i) {
+      if (!isa<ArraySubscriptExpr>(MemberExprBases[i]))
+        break;
+      ++Dims;
+    }
+    return Dims;
+  }
```

Comment:
Can we just save array nesting depth and increment/decrement it once we enter/leave array?

Reply:
I don't think we can just save that, since a sum of those ends up getting confused if there is a struct in the middle.

Reply:
Yep, exactly.  We'd have to make sure we 'reset' that counter (or stashed it somewhere?) when we enter/leave structs.  

In reality, we probably could get a benefit from keeping a stack of things we have 'entered' (once the foolish 'enter field goes away), but this is likely what we can do for now.

Reply:
The recovery of array indices would be simplified by maintaining a stack of index values explicitly. Early in the implementation of arrays such a stack did exist, but was rejected as being unnecessary. Strictly speaking, It is unnecessary as this code proves, but this method of recovery of the indices from the MemberExpr stack is messy.

Reply:
> but this method of recovery of the indices from the MemberExpr stack is messy.

Yes it is. But I hope we manage to refactor it later.

Reply:
Yep. Hopefully Erich's changes to the model, simplifies this.